anvil/src/state.rs

```diff
@@ -418,14 +436,55 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
 
                 send_throttled_frame(surface, states, report, now, throttle)
             });
+
+            window.send_presentation_feedback(output, tv_sec, tv_nsec, refresh, seq, |surface, _| {
+                let is_on_primary_output = window_surface_primary_output(surface, &*outputs)
+                    .map(|primary_output| primary_output == output)
+                    .unwrap_or(false);
+
+                if !is_on_primary_output {
+                    return None;
+                }
+
+                report.element_render_state(surface).map(|state| {
+                    let mut flags = wp_presentation_feedback::Kind::Vsync;
+                    if state == RenderElementState::ScannedOut {
+                        flags |= wp_presentation_feedback::Kind::ZeroCopy;
+                    }
+                    flags
+                })
+            });
         }
 
         let map = smithay::desktop::layer_map_for_output(output);
         for layer_surface in map.layers() {
             layer_surface.send_frame(time, |surface, states| {
                 send_throttled_frame(surface, states, report, now, throttle)
             });
+
+            layer_surface.send_presentation_feedback(output, tv_sec, tv_nsec, refresh, seq, |surface, _| {
+                report.element_render_state(surface).map(|state| {
+                    let mut flags = wp_presentation_feedback::Kind::Vsync;
+                    if state == RenderElementState::ScannedOut {
+                        flags |= wp_presentation_feedback::Kind::ZeroCopy;
+                    }
+                    flags
+                })
+            })
+        }
+    }
+}
+
+pub fn clock_get_time(clk_id: libc::clockid_t) -> Result<libc::timespec, std::io::Error> {
```

Comment:
Do we need this function? `std::time::Instant` is defined as using `clock_gettime` with the monotonic clock.
https://doc.rust-lang.org/nightly/std/time/struct.Instant.html#underlying-system-calls

Reply:
Yeah, but we need tv_sec and tv_nsec for wp_presentation_feedback and I don't see how we could get that from Instant without some hacks (looking at you std::mem::zeroed).
The client has to be able to use the same clk_id to calculate time diffs, so we need to use the same base AFAIK.

Reply:
Yeah, this is quite annoying. `std::time::Instant` really needs some function, that allow for C-interop. Preferably just a `Instant::ZERO` to get a zero-base, but that probably doesn't mean much on other systems.

Reply:
I don't think this would be helpful, the intended use of `Instant` is just not well suited for our use-case. The fact that it uses `timespec` under the hood is just an implementation detail. Beside that `Instant` explicitely does not define a base so it is unsuitable for transporting a timepoint which requires a known base.